crates/sui-indexer-alt-graphql/src/api/query.rs

```diff
@@ -555,27 +568,52 @@ impl Query {
 
     /// Simulate a transaction to preview its effects without executing it on chain.
     ///
-    /// - `transactionDataBcs` contains the BCS-encoded transaction data (Base64-encoded).
+    /// Accepts either BCS-encoded transaction data or a JSON representation that can be partially resolved.
     ///
     /// Unlike `executeTransaction`, this does not require signatures since the transaction is not committed to the blockchain. This allows for previewing transaction effects, estimating gas costs, and testing transaction logic without spending gas or requiring valid signatures.
     async fn simulate_transaction(
         &self,
         ctx: &Context<'_>,
-        transaction_data_bcs: Base64,
+        data: SimulateTransactionData,
     ) -> Result<SimulationResult, RpcError<TransactionInputError>> {
         let fullnode_client: &FullnodeClient = ctx.data()?;
 
-        // Parse transaction data from BCS
-        let tx_data: TransactionData = {
-            let bytes: &Vec<u8> = &transaction_data_bcs.0;
-            bcs::from_bytes(bytes)
-                .map_err(|err| bad_user_input(TransactionInputError::InvalidTransactionBcs(err)))?
+        let proto_tx = match data {
+            SimulateTransactionData::Bcs(bcs_data) => {
+                // Parse BCS to get TransactionData first
+                let bytes: &Vec<u8> = &bcs_data.0;
+                let tx_data: TransactionData = bcs::from_bytes(bytes).map_err(|err| {
+                    bad_user_input(TransactionInputError::InvalidTransactionBcs(err))
+                })?;
+
+                // Create proto with BCS
+                let mut proto = proto::Transaction::default();
+                proto.bcs = Some(proto::Bcs::serialize(&tx_data).map_err(|err| {
+                    bad_user_input(TransactionInputError::InvalidTransactionBcs(err))
+                })?);
+                proto
+            }
+            SimulateTransactionData::Json(json_str) => {
+                // Parse JSON directly to proto::Transaction
+                serde_json::from_str(&json_str).map_err(|err| {
```

Comment:
@bmwill, would using `serde_json` to deserialize into a `proto` type definition understand the canonical representation of the proto structure? I was thinking that in some cases (e.g. binary data) the two approaches would differ, and that in those cases we should prefer the canonical proto representation.

Reply:
Okay nice -- so that means that `{ bcs: "<Base64, BCS-encoded TransactionData>" }` would deserialize correctly, right?

Reply:
For example this is looks like when querying a random txn from the chain:
```
"bcs": {
                        "name": "TransactionData",
                        "value": "AAl5AwAAAAAAALifAwAAAAAAABmssFmZAQAAICutV38G0ElwfXYaCntLk0C2Enyla8PuCcFf/vI2gk/8AQAgsI8So1XIm+P+/64o6MX3eLAxv2V8UzH+unNE4G5lDD0AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAAAAAAAAA=="
                    },
```

The `name` field is extra information and may or may not be required when providing Bcs as input (its not required for simulate/execute because we expect `TransactionData`